File: app/llm/prompts.py
"""AI reviewer prompt (BUILD_PLAN.md §5), as a constant with a version tag —
the same tag gets written into every export row's metadata in P6 (§7.5)."""
import json
from typing import Any, Optional

REVIEWER_PROMPT_VERSION = "v1"

REVIEWER_SYSTEM_PROMPT = (
    "You are an investment thesis auditor. You are given a written thesis and the "
    "latest verified operating data. Decide whether the thesis is intact. You must "
    "reason only from the data provided; if the data does not settle a question, say "
    "so and lower your confidence rather than assuming. Return JSON only."
)


def build_reviewer_user_prompt(
    company_name: str,
    broad_industry: str,
    specific_niche: str,
    operating_model: str,
    authored_at: str,
    thesis_data: dict[str, Any],
    period: str,
    metrics: list[dict],
    narrative: Optional[str],
    rule_engine_findings: list[dict],
) -> str:
    redlines = [t.get("label") for t in thesis_data.get("what_can_kill_it", [])]
    return (
        f"COMPANY: {company_name} | {broad_industry} > {specific_niche} | {operating_model}\n"
        f"THESIS (as written on {authored_at}):\n"
        f"  The Business: {thesis_data.get('the_business', {}).get('what_it_does', '')}\n"
        f"  The Big Change: {thesis_data.get('the_big_change', {}).get('summary', '')}\n"
        f"  Why We Believe It: {thesis_data.get('why_we_believe_it', [])}\n"
        f"  Invalidation Redlines: {redlines}\n"
        f"NEW EVIDENCE ({period}):\n"
        f"  metrics: {json.dumps(metrics)}\n"
        f"  narrative: {narrative or ''}\n"
        f"RULE ENGINE FINDINGS: {json.dumps(rule_engine_findings)}\n\n"
        "Return: {\"verdict\": \"on_track|watch_closely|broken\", \"confidence\": 0.0-1.0, "
        "\"reasoning_chain\": [\"Premise 1: ...\",\"Premise 2: ...\",\"Inference: ...\",\"Conclusion: ...\"], "
        "\"evidence_used\": [\"metric_key\", ...], \"unresolved_questions\": [\"...\"]}"
    )

File: web/src/lib/server/llm/prompts.ts
// Ports app/llm/prompts.py. Same version tag gets written into every export
// row's metadata (see exporter.ts) - keep this string in lockstep with the
// Python REVIEWER_PROMPT_VERSION if the prompt ever changes.
export const REVIEWER_PROMPT_VERSION = 'v1';

export const REVIEWER_SYSTEM_PROMPT =
	'You are an investment thesis auditor. You are given a written thesis and the ' +
	'latest verified operating data. Decide whether the thesis is intact. You must ' +
	'reason only from the data provided; if the data does not settle a question, say ' +
	'so and lower your confidence rather than assuming. Return JSON only.';

interface ThesisDataLike {
	the_business?: { what_it_does?: string };
	the_big_change?: { summary?: string };
	why_we_believe_it?: string[];
	what_can_kill_it?: Array<{ label?: string }>;
}

export function buildReviewerUserPrompt(opts: {
	companyName: string;
	broadIndustry: string;
	specificNiche: string;
	operatingModel: string;
	authoredAt: string;
	thesisData: ThesisDataLike;
	period: string;
	metrics: unknown[];
	narrative: string | null | undefined;
	ruleEngineFindings: unknown[];
}): string {
	const { companyName, broadIndustry, specificNiche, operatingModel, authoredAt, thesisData, period, metrics, narrative, ruleEngineFindings } =
		opts;
	const redlines = (thesisData.what_can_kill_it ?? []).map((t) => t.label);
	return (
		`COMPANY: ${companyName} | ${broadIndustry} > ${specificNiche} | ${operatingModel}\n` +
		`THESIS (as written on ${authoredAt}):\n` +
		`  The Business: ${thesisData.the_business?.what_it_does ?? ''}\n` +
		`  The Big Change: ${thesisData.the_big_change?.summary ?? ''}\n` +
		`  Why We Believe It: ${JSON.stringify(thesisData.why_we_believe_it ?? [])}\n` +
		`  Invalidation Redlines: ${JSON.stringify(redlines)}\n` +
		`NEW EVIDENCE (${period}):\n` +
		`  metrics: ${JSON.stringify(metrics)}\n` +
		`  narrative: ${narrative || ''}\n` +
		`RULE ENGINE FINDINGS: ${JSON.stringify(ruleEngineFindings)}\n\n` +
		'Return: {"verdict": "on_track|watch_closely|broken", "confidence": 0.0-1.0, ' +
		'"reasoning_chain": ["Premise 1: ...","Premise 2: ...","Inference: ...","Conclusion: ..."], ' +
		'"evidence_used": ["metric_key", ...], "unresolved_questions": ["..."]}'
	);
}
